grouper/repositories/user.py

```diff
@@ -1,29 +1,97 @@
+from collections import defaultdict
 from typing import TYPE_CHECKING
 
-from grouper.entities.user import UserNotFoundException
-from grouper.models.user import User
+from grouper.entities.user import PublicKey, User, UserMetadata, UserNotFoundException
+from grouper.models.public_key import PublicKey as SQLPublicKey
+from grouper.models.user import User as SQLUser
+from grouper.models.user_metadata import UserMetadata as SQLUserMetadata
+from grouper.repositories.interfaces import UserRepository
 
 if TYPE_CHECKING:
+    from grouper.graph import GroupGraph
     from grouper.models.base.session import Session
+    from typing import Dict, List
 
 
-class UserRepository(object):
-    """Storage layer for users."""
+class GraphUserRepository(UserRepository):
+    """Graph-aware storage layer for users."""
+
+    def __init__(self, graph, repository):
+        # type: (GroupGraph, UserRepository) -> None
+        self.graph = graph
+        self.repository = repository
+
+    def all_enabled_users(self):
+        # type: () -> Dict[str, User]
+        return self.graph.all_user_metadata()
+
+    def disable_user(self, user):
+        # type: (str) -> None
+        self.repository.disable_user(user)
+
+    def user_is_enabled(self, name):
+        # type: (str) -> bool
+        """Return whether a user is enabled.
+
+        TODO(rra): This checks the underlying data store, not the graph, even though this
+        information is in the graph, because the convert_user_to_service_account usecase disables a
+        user and then immediately checks whether the user is disabled, and we don't want to force a
+        graph refresh in the middle of that usecase.  This indicates a deeper underlying problem
+        where some usecases need to use SQL repositories rather than the graph, which we're
```

Comment:
The problem is that the graph is a cache and the assumption is that a graph repository will answer the questions that it can from the cache, but here we're not answering a question from the graph even though it's in the graph because it would break another use case. That in turn points to a deeper problem that the graph becomes inconsistent on writes, and we don't have a way of resolving that other than forcing a graph update, which is slow and is much of why the Grouper frontend is slow.

I think in an ideal world (and this also answers your comment below), APIs that allow writes would not use the graph at all (for either read or write), and the graph would be for read-only use cases, thus avoiding the inconsistency problem.